### PR TITLE
Reload mules and spoolers

### PR DESCRIPTION
**Pull request by Shir0kamii**
Hello,

This PR relates to issue #915.

We didn't really need the mules and spooler to be chain reloaded, so I thought a general option would be better for other users of uwsgi.

I added two new options to enable the reloading of the mules and the spoolers upon the modification of a file. It's independent from the reloading of the workers.

What do you think ? Is it good for you ?

### CODE REVIEW

**Comment by xrmx** [comment]
@Shir0kamii tabs please :)

**Comment by Shir0kamii** [comment]
@xrmx Sorry, that's good now!

File: core/master.c [hunk 1]
```diff
@@ -861,6 +870,18 @@ int master_loop(char **argv, char **environ) {
...
+				touched = uwsgi_check_touches(uwsgi.touch_mules_reload);
+				if (touched) {
+					uwsgi_log_verbose("*** %s has been touched... mules reload !!! ***\n", touched);
+					uwsgi_reload_mules();
+					continue;
+				}
+				touched = uwsgi_check_touches(uwsgi.touch_spoolers_reload);
+				if (touched) {
+					uwsgi_log_verbose("*** %s has been touched... spoolers reload !!! ***\n", touched);
+					uwsgi_reload_spoolers(SIGHUP);
```
Comment (unbit) [inline core/master.c]: why the SIGHUP arg here ?

**Comment by unbit** [comment]
The patch looks really good, i know i am asking a bit too much, but do you think is it possible to add a test for it like i did it here ?

https://github.com/unbit/uwsgi/commit/aeb2fc4f883da7d5b8d5c3fb88bcd74f50ce987b

it is pretty hard to test process management, but maybe it is enough to get infos from the stats server to check if the pid of the mules/spoolers change at the right time.

**Comment by Shir0kamii** [comment]
Sorry, the SIGHUP was a mistake.

I will try to test these two features and come back to you soon!

**Comment by Shir0kamii** [comment]
@unbit I did my best, tell me what you think of these tests

I didn't use the uwsgi stats server since it doesn't send any data about the mules (though it does for the spooler). Instead, I test the list of subprocesses of uwsgi before and after reloading.

the sleeps pains me but if didn't put them here, uwsgi wouldn't have time to reload and the test fails.

If you have a better idea to test this feature, just tell me!

**Comment by unbit** [comment]
the approach is good, maybe i would enable the --auto-procname feature to recognize spooler and mules processes by the name

**Comment by Shir0kamii** [comment]
I didn't know about that feature until now, it seems great !

I see it renames the processes of uwsgi, but I don't understand how you would use it in the context of the tests. Could you explain your thoughts please ?

**Comment by xrmx** [comment]
Il 27/07/2016 19:31, Alexandre Bonnetain ha scritto:
> I didn't know about that feature until now, it seems great !
>
> I see it renames the processes of uwsgi, but I don't understand how you would
> use it in the context of the tests. Could you explain your thoughts please ?

So you can grep for the spooler / mules and see if that process has changed pid 
instead of checking the uwsgi one.

**Comment by Shir0kamii** [comment]
@xrmx @unbit It seems I have learnt a lot about process management while making these tests, thank you for your guidance!

They should now be just like you want, tell me if something still bother you

File: uwsgi.h [hunk 2]
```diff
@@ -2361,6 +2364,8 @@ struct uwsgi_server {
 
 	struct uwsgi_string_list *touch_reload;
 	struct uwsgi_string_list *touch_chain_reload;
+	struct uwsgi_string_list *touch_mules_reload;
```
Comment (unbit) [inline uwsgi.h]: please move new structure fields at the and, to maintain ABI compatibility

**Comment by Shir0kamii** [comment]
@unbit This is done! Is there anything left before merging ?

**Comment by unbit** [comment]
@xrmx safe for backport too

**Comment by ultrabug** [comment]
Great news thanks @Shir0kamii and @unbit @xrmx !

I hope this gets tagged soon so we can have a real package for this :+1:

**Comment by xrmx** [comment]
Backported to uwsgi-2.0 as 24fd91e1000251fa44f0cfd630bcf3c867336264

**Comment by ultrabug** [comment]
Thanks, any planned 2.0.14 release plz ?

**Pull request merged**